### PR TITLE
Fix serialization issues, add automatic check

### PR DESCRIPTION
**Pull request by msperber**
This addresses #368:
- adds an automatic check and raises an exception when the components written out don't include all of the allocated DyNet parameters. The exception will be raised before the training loop is entered.
- fixes LinearBridge, as well as a couple of other places that had the same problem.

The transformer test currently fails because of this new check, so I marked it as ```@unittest.expectedFailure```. The reason it fails is that it uses a lot of nested classes that are all not Serializable even though they should be, and then allocates Serializable classes at the bottom of the hierarchy without using the ```add_serializable_component``` idiom. It might be worth addressing this later, as part of #320?

### CODE REVIEW

**Comment by msperber** [comment]
Strange, we're suddenly getting a segmentation fault when calling loss.backward(). This is not happening when running tests locally, and this PR is doing nothing I can think of that would cause this behavior..

**Comment by neubig** [comment]
Regarding the transformer, I'll be doing some refactoring on this to make it more canonical xnmt style, so don't worry about it for now.

**Comment by msperber** [comment]
Tests are passing now, so this is ready to merge. FYI, the segmentation fault occurred only with the pip version of DyNet and not with the version from the master branch. The workaround was to revert some unimportant and seemingly unrelated changes. I won't investigate further because it seems this is a DyNet bug that has been fixed.
